core/trino-main/src/main/java/io/trino/metadata/NodeState.java

```diff
@@ -16,6 +16,8 @@
 public enum NodeState
 {
     ACTIVE,
+    DECOMMISSIONED,
+    DECOMMISSIONING,
```

Comment:
High level question. why do we need a concept of "decommissioned" node? Why can't we just kill nodes to be scaled down. And if we scale up, add new ones?

Reply:
A node can stay in graceful shutdown state for a long time potentially while it is draining running tasks. During this time scaling requirements can change. Having the ability to bring back nodes from graceful shutdown state to active can be useful for getting nodes immediately rather than waiting to acquire containers/nodes, bring up Trino server process and have it join the cluster. It also potentially saves resources as we might have a situation of many nodes in graceful shutdown which don't do much work but we bring up new nodes because we can't "recommission" them.

Reply:
DECOMMISSIONED is a state of the Trino worker. Node (in our case the k8s pod) termination usually happens after. Once the worker is DECOMMISSIONED, the node/pod can be terminated gracefully without impact to any query. Basically DECOMMISSIONED is a logical state where no task will be scheduled on the worker, despite the TrinoServer process is still running.

In majority cases, the pod is terminated afterward but in rare cases, a DECOMMISSIONED node can be recommissioned, if the query traffic justifies, so we save the termination of a pod and allocation of a new one.

Reply:
@raunaqmorarka Just refreshed and noticed your comment. Your comment is on the point and I agree.

Reply:
Ok - `DECOMISSIONING` state indeed makes sense. Not sure about `DECOMISSINED` - as starting up new Trino process should be negilibible cost in most scenrios. But maybe it may be useful. I guess handling `DECOMISSIONING` and `DECOMISSIONED` is same anyway.

Reply:
So the other question then? Maybe given the fact that we have fault-tolerant execution mode right now this is not really a crucial feature. With FTE you can just shoot the node, not wait for a graceful shutdown at all, or just wait for a limited time. 
Some work will be wasted, but the query as a whole will survive. Would that be feasible for your use-case @dzhi-lyft ? 

@arhimondr any comments on this one?

Reply:
Task retry on failures (FTE) would increase the robustness of Trino cluster in cases like node failure (unexpected termination, persistent error) etc. It is more of an independent feature as to node-level graceful decommission. The key flexibility this PR provides is bi-directional transition between ACTIVE and DECOMMISSIONING state so to reduce regret during auto-scaling. 

We have autoscaler component periodically (every 40 seconds) polls status of the clusters and decides the desired node count based on recent cluster utilization, such desired count may change due to query traffic fluctuation (particularly in clusters where the query time out is 2 hours), and some nodes may transition from DECOMMISSIONING back to ACTIVE as needed. It avoids unnecessary termination of pods which sometime wait in their pre-stop hook for long time, and avoids reconciliation confusion in k8s controller with large number of pods in TERMINATING and new ones are requested.

Reply:
> It avoids unnecessary termination of pods which sometime wait in their pre-stop hook for long time

Sounds like this is the source of the problem?

Is your intention to revert the scale down decision while nodes being terminated are still being drained?